### PR TITLE
topdown: Ignore errors during parsing Expires header

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
Earlier if the value for the Expires header was set
to an invalid date like 0, we would return an error and
fallback to the normal cache (ie.we would not use
the inter-query cache).

Servers can sometimes set `Expires: 0` to indicate
expired content. This change ignores the errors
resulting from parsing the Expires header.

Fixes: #3284

Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

**Review by anderseknert — APPROVED** [review]
LGTM!

With regards to the issue reported; one thing we might want to consider is why we even parse the date/time headers at all when the client explicitly told us to ignore them? We could probably short-circuit evaluation earlier when the force cache directives are present. But fixing the `Expires: 0` case is good regardless of that.

File: topdown/http.go [hunk 1]
```diff
@@ -936,15 +933,22 @@ func getResponseHeaderDate(headers http.Header) (date time.Time, err error) {
...
 	expiresHeader := headers.Get("expires")
 	if expiresHeader == "" {
-		return
+		return time.Time{}
 	}
-	return time.Parse(time.RFC1123, expiresHeader)
+
+	date, err := http.ParseTime(expiresHeader)
+	if err != nil {
+		// servers can set `Expires: 0` which is an invalid date to indicate expired content
```
Comment (srenatus) [inline topdown/http.go:947]: I suppose we could have checked explicitly if the expires header is "0" instead? The way it is done now, we could ignore errors when it's neither valid nor "0".... 🤔 

(Just checked, that's what @anderseknert has proposed, too... 😄)

Comment (anderseknert) [reply topdown/http.go:947]: The [spec](https://tools.ietf.org/html/rfc7234#section-5.3) says:
> A cache recipient MUST interpret invalid date formats, especially the value "0", as representing a time in the past (i.e., "already expired").

So while it does call out "0" especially, it seems fine to treat any other invalid dates the same way.

What I proposed was rather that we don't check these headers _at all_ when the client is configured with `force_cache: true` + `force_cache_duration` 😊

Comment (srenatus) [reply topdown/http.go:947]: Ah I see. Thanks for digging out the spec! 🔍 

Sounds like we should get this merged then :D

**Comment by ashutosh-narkar** [comment]
> With regards to the issue reported; one thing we might want to consider is why we even parse the date/time headers at all when the client explicitly told us to ignore them? We could probably short-circuit evaluation earlier when the force cache directives are present. But fixing the `Expires: 0` case is good regardless of that.

That's a good point @anderseknert. We don't need to parse all the headers when force cache is.

**Pull request merged**